test.py

```diff
@@ -0,0 +1,117 @@
+    ap.add_argument('--opal_num_gru_layers', type=int, default=4, help='OPAL: number of GRU layers')
+    ap.add_argument('--opal_state_agnostic', action='store_true', help='OPAL: if set, use state agnostic models')
+    ap.add_argument('--opal_unit_prior_std', action='store_true', help='OPAL: if set, use fixed std=1 for prior')
+
+    ap.add_argument('--hidden_size', type=int, default=256, help='size of hidden layers (H)')
+    ap.add_argument('--num_layers', type=int, default=3, help='number of hidden layers')
+    ap.add_argument('--num_gru_layers', type=int, default=4, help='number of GRU layers')
+
+    # downstream task
+    ap.add_argument('--task_type', type=str, choices=['offline', 'imitation', 'online', 'multitask'], default='offline', help='downstream task type')
+    ap.add_argument('--policy_type', type=str, choices=['cql', 'bc', 'sac', 'ppo'], default='cql', help='RL algorithm to use for downstream learning')
+    ap.add_argument('--batch_size', type=int, default=50, help='batch size')
+    ap.add_argument('--num_workers', type=int, default=6, help='number of DataLoader workers')
+
+    args = ap.parse_args()
+
+    for attr in ['opal_ckpt', 'task_ckpt']:
+        val = getattr(args, attr)
+        setattr(args, attr, os.path.expanduser(val) if val else val)
+
+    return args
+
+
+def test(args):
+    device = python_utils.get_device(args.gpu_id)
+    python_utils.seed_all(args.seed)
+    multitask = args.task_type == 'multitask'
+    env = env_utils.make_env(args.domain_name, args.task_name, seed=args.seed, eval=True,
+                             sparse_reward=True, append_task=multitask)
+
+    dim_s = np.prod(env.observation_space.shape)
+    dim_s_opal = dim_s - np.prod(env.task_space.shape) if multitask else dim_s
+    dim_a = np.prod(env.action_space.shape)
+    dim_z = args.latent_dim
+
+    # load trained OPAL
+    opal = OPAL(
+        dim_s=dim_s_opal,
+        dim_a=dim_a,
+        dim_z=dim_z,
+        hidden_size=args.opal_hidden_size,
+        num_layers=args.opal_num_layers,
+        num_gru_layers=args.opal_num_gru_layers,
+        state_agnostic=args.opal_state_agnostic,
+        unit_prior_std=args.opal_unit_prior_std,
+    ).to(device)
+    opal.load_state_dict(torch.load(args.opal_ckpt))
+
+    # load trained task model
+    model = HRLAgent(args.policy_type, opal, longstep_len=args.subtraj_len, multitask=multitask,
+                     dim_s=dim_s, dim_z=dim_z,
+                     hidden_size=args.hidden_size, num_layers=args.num_layers).to(device)
+    model.load_state_dict(torch.load(args.task_ckpt))
+
+    # rollout
+    returns = []
+    successes = []
+    ep_lengths = []
+
+    for ep_num in range(1, args.test_episodes + 1):
+        tic = time.time()
+        obs, ret, done, success = env.reset(), 0., False, False
+        step, ep_len = 0, 0
+
+        while not done:
+            next_step, next_obs, r, done, success, *_ = model.rollout_step(env, step, obs, deterministic=True)
```

Comment:
Q) What does `*_` do?

Reply:
It compresses all the trailing return values into a variable `_` (which would become a tuple). Since `rollout_step` returns a tuple with more elements than I need, I'm truncating all the elements after `success`.